verification/spec/relayer/ClientHandlers.tla

```diff
@@ -0,0 +1,106 @@
+ Client helper operators
+ ***************************************************************************)
+
+\* get the ID of chainID's counterparty chain    
+GetCounterpartyChainID(chainID) ==
+    IF chainID = "chainA" THEN "chainB" ELSE "chainA"    
+ 
+\* get the client ID of the client for chainID 
+GetClientID(chainID) ==
+    IF chainID = "chainA" THEN "clA" ELSE "clB"
+        
+\* get the client ID of the client for chainID's counterparty chain           
+GetCounterpartyClientID(chainID) ==
+    IF chainID = "chainA" THEN "clB" ELSE "clA" 
+    
+
+(***************************************************************************
+ Client datagram handlers
+ ***************************************************************************)
+    
+\* client heights: 
+\* good version: add all client heights from datagrams to counterpartyClientHeights
+\* bad version: add only Max of client heights from datagrams to counterpartyClientHeights
+\*              if Max > Max(counterpartyClientHeights)
+    
+\* Handle "CreateClient" datagrams
+HandleCreateClient(chainID, chain, datagrams) == 
+    \* get "CreateClient" datagrams with valid clientID
+    LET createClientDgrs == {dgr \in datagrams : 
+                            /\ dgr.type = "CreateClient"
+                            /\ dgr.clientID = GetCounterpartyClientID(chainID)} IN
+    \* get heights in datagrams with correct counterparty clientID for chainID
+    LET createClientHeights == {h \in Heights : \E dgr \in createClientDgrs : dgr.height = h} IN  
+    
+    \* new chain record with clients created
+    LET clientCreateChain == [
+            height |-> chain.height,
+            counterpartyClientHeights |-> 
+                \* if set of counterparty client heights is not empty
+                IF chain.counterpartyClientHeights /= {}
+                \* then discard CreateClient datagrams  
+                THEN chain.counterpartyClientHeights
+                \* otherwise, if set of heights from datagrams is not empty  
+                ELSE IF createClientHeights /= {} 
+                     \* then create counterparty client with height Max(createClientHeights) 
+                     THEN {Max(createClientHeights)}
+                     \* otherwise, do not create client (as chain.counterpartyClientHeight = {})  
+                     ELSE chain.counterpartyClientHeights,
+            connectionEnd |-> chain.connectionEnd
+         ] IN
+
+   clientCreateChain
+ 
+\* Handle "ClientUpdate" datagrams
+HandleUpdateClient(chainID, chain, datagrams) ==      
+    \* get "ClientUpdate" datagrams with valid clientID
+    LET updateClientDgrs == {dgr \in datagrams : 
+                            /\ dgr.type = "ClientUpdate"
+                            /\ dgr.clientID = GetCounterpartyClientID(chainID)} IN
+    \* get heights in datagrams with correct counterparty clientID for chainID
+    LET updateClientHeights == {h \in Heights : \E dgr \in updateClientDgrs : dgr.height = h} IN    
```

Comment:
Seems that here we don't know the order of the client updates and we update with all heights in the datagrams. If we want to catch the current problem with the chain (failing to update the client state with a lower height than the max it has) then we need to consider sequence instead of sets (?)

Reply:
As discussed, we will do these changes as part of multiple relayer version.